src/Processors/QueryPlan/Optimizations/removeRedundantDistinct.cpp

```diff
@@ -0,0 +1,192 @@
+#include <memory>
+#include <Interpreters/ActionsDAG.h>
+#include <Processors/QueryPlan/ArrayJoinStep.h>
+#include <Processors/QueryPlan/DistinctStep.h>
+#include <Processors/QueryPlan/ExpressionStep.h>
+#include <Processors/QueryPlan/FillingStep.h>
+#include <Processors/QueryPlan/FilterStep.h>
+#include <Processors/QueryPlan/IntersectOrExceptStep.h>
+#include <Processors/QueryPlan/JoinStep.h>
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+#include <Processors/QueryPlan/UnionStep.h>
+#include <Common/logger_useful.h>
+#include <Common/typeid_cast.h>
+
+namespace DB::QueryPlanOptimizations
+{
+
+namespace
+{
+    constexpr bool debug_logging_enabled = false;
+    void logActionsDAG(const String & prefix, const ActionsDAGPtr & actions)
+    {
+        if constexpr (debug_logging_enabled)
+        {
+            LOG_DEBUG(&Poco::Logger::get("redundantDistinct"), "{}: {}", prefix, actions->dumpDAG());
+        }
+    }
+
+    std::set<std::string_view> getDistinctColumns(const DistinctStep * distinct)
+    {
+        /// find non-const columns in DISTINCT
+        const ColumnsWithTypeAndName & distinct_columns = distinct->getOutputStream().header.getColumnsWithTypeAndName();
```

Comment:
Now I see that there is a list of distinct columns in `DistinctStep`
https://github.com/ClickHouse/ClickHouse/blob/03adbfdef773f43b192acd580f227cf1f7d9815e/src/Processors/QueryPlan/DistinctStep.h#L36

Let's use only columns from it just in case. (I hope it's same as header, but still)